rmw_opendds_shared_cpp/src/init.cpp

```diff
@@ -13,17 +13,68 @@
 // limitations under the License.
 
 #include "rmw_opendds_shared_cpp/init.hpp"
+#include "rmw_opendds_shared_cpp/RmwAllocateFree.hpp"
+#include "rmw/init_options.h"
 #include "rmw/error_handling.h"
 #include "dds/DCPS/Service_Participant.h"
 
+rmw_context_impl_t::rmw_context_impl_t() : dpf_(TheParticipantFactory)
+{
+  if (dpf_) {
+    TheServiceParticipant->set_default_discovery(OpenDDS::DCPS::Discovery::DEFAULT_RTPS);
+  } else {
+    const char* msg = "failed to get participant factory";
+    RMW_SET_ERROR_MSG(msg);
+    throw std::runtime_error(msg);
+  }
+}
+
+rmw_context_impl_t::~rmw_context_impl_t()
+{
+  if (dpf_) {
+    TheServiceParticipant->shutdown();
+    dpf_ = nullptr;
+  }
+}
+
 rmw_ret_t
-init()
+init(rmw_context_t& context)
 {
-  DDS::DomainParticipantFactory * dpf_ = TheParticipantFactory;
-  if (!dpf_) {
-    RMW_SET_ERROR_MSG("failed to get participant factory");
-    return RMW_RET_ERROR;
+  context.impl = RmwAllocateFree<rmw_context_impl_t>::create();
+  if (context.impl) {
+    return RMW_RET_OK;
+  }
+  context = rmw_get_zero_initialized_context();
+  return RMW_RET_ERROR;
+}
+
+// See rmw/init.h: rmw_ret_t rmw_shutdown(rmw_context_t * context)
+// "If context has been already invalidated (`rmw_shutdown()` was called on it),
+// then this function is a no-op and `RMW_RET_OK` is returned."
+rmw_ret_t
+shutdown(rmw_context_t& context)
+{
+  if (context.impl) {
+    RmwAllocateFree<rmw_context_impl_t>::destroy(context.impl);
   }
-  TheServiceParticipant->set_default_discovery(OpenDDS::DCPS::Discovery::DEFAULT_RTPS);
   return RMW_RET_OK;
 }
+
+bool is_zero_initialized(const rmw_init_options_t * o)
+{
+  return o && o->instance_id == 0
+    && !(o->implementation_identifier)
+    && o->domain_id == RMW_DEFAULT_DOMAIN_ID
+    //?? RMW_SECURITY_ENFORCEMENT_PERMISSIVE in rmw_get_default_security_options()
+    && o->security_options.enforce_security == 0 && !(o->security_options.security_root_path)
+    && o->localhost_only == RMW_LOCALHOST_ONLY_DEFAULT
+    && !(o->enclave)
+    && !(o->impl);
+}
+
+bool is_zero_initialized(const rmw_context_t * c)
```

Comment:
These are pretty short functions so it's less important to use meaningful variable names, but I feel like it's always better practice to use a variable name that shows intent rather than a single character. It's also harder to search a single character variable name in a broader context like a workspace.

Reply:
done